### DISCUSSION TITLE
【求助】几个小需求还望答疑

### DISCUSSION THREAD

**Discussion by lhqm** [simple-robot/Discussions#145] [Q&A] [answered]
### 涉及的编程语言

Java

### 涉及的组件库

mirai (QQ)

#### 0.前置说明
项目已加Redis、MySQL和influxdb，需要涉及到相关库操作的可以直接提供逻辑，不做详细讲解，感激不尽
#### 1.关于延时处理事件
比如说有一个添加好友的请求或者邀请加群的请求被我捕获到了，但是这个请求应该让bot管理员手动确认是否同意，然后再由bot去执行相关代码逻辑，但是这个过程往往可能长达数个小时，这应该怎么处理呢？
#### 2.关于图片和文件的下载和上载
这个能给个文档指路嘛，萌新实在没找到，QAQ
#### 3.关于自定义过滤器
比如说我有一个黑名单设置，黑名单内的请求是应该被快速失败的，那么我们有提供非切面写法的过滤器来进行快速失败嘛？以及是否能在某个节点将请求直接失败返回，防止之后可能存在的同类型监听器还想去执行代码呢？
#### 4.关于监听器优先级
监听器优先级能否被设置呢？比如说我想让消息事件先走监听器1，再走监听器2

**Comment by ForteScarlet** [comment, marked as answer]
1. 关于延时处理事件

    延迟方案有很多，比较简单的例子有使用延迟队列或使用定时任务定期检测缓存中的待处理内容。

2. 关于图片和文件的下载和上载

    - 图片发送可参考 #115  , #92  等
    - 图片下载可参考 #15  , 以及 `Image` 中的 `resource` 属性的各实现类。（例如，假如判断类型为 `URLResource` ，则可以直接获取 `url` 而不必使用数据流下载图片）

3. 关于自定义过滤器

    - 首先最简单的办法即参考 #62 ，并提供一个最高优先级的、监听所有事件类型的监听器，这样就能算一种“全局拦截器”了。
    - 其次，也可以考虑使用 _拦截器_ ，在Java中的话，选择 [BlockingEventListenerInterceptor](https://docs.simbot.forte.love/main/simbot-api/love.forte.simbot.event/-blocking-event-listener-interceptor/index.html) 或 [BlockingEventProcessingInterceptor](https://docs.simbot.forte.love/main/simbot-api/love.forte.simbot.event/-blocking-event-processing-interceptor/index.html) 实现并注册至Spring容器中即可。但是 `Blocking` 拦截器会直接影响到被拦截所有目标使其变为阻塞形式，影响未知（如果全Java也许也没什么大碍吧，没详细测试过）。

4. 关于监听器优先级

    - 参考 #62 
    - 优先级属性在 `@Listener` 中

**Reply by lhqm** [reply]
谢谢大佬的耐心解答！感激不尽！

**Reply by lhqm** [reply]
不过这个延时处理的话大佬可以给一段参考嘛？萌新还是不太懂应该如何从缓存里边拉到相关事件数据，是直接通过事件的编号来进行new对象行为进行处理，或者从springboot context获取相关缓存嘛？还是怎么样哇？

**Reply by ForteScarlet** [reply]
最简单的办法之一，用一个 `Map` 缓存一下就行，Key可以是事件ID，如果不放心也可以自己随机一个 `UUID` 。
